Make using alloc optional and usable on stable Rust

sirhcel opened the pull request:
Thank you for this awesome crate Andrei! I'd really like the using the formatters from PrettyHex::hex_dump which feels pretty elegant for logging. :)

The previously used 'feature(alloc)' got rejected on stable Rust 1.46.0. By switching to 'cfg(feature = "alloc")' this crate builts on stable and with the Cargo configuration 'default-features = false' with a reduced interface even without an allocator.

I'm not firm with nightly Rust and don't now much about `#[feature(...)]`. Does https://github.com/sirhcel/pretty-hex/commit/1a5542ac468ce3e02b66acfed1fbf5d541db7537#diff-b4aea3e418ccdb71239b96952d9cddb6L2 enable the stuff depending on String for the whole library? To me it looks like using `alloc` is generally enabled by default. So I explicitly turned it on in my change. Please feel free to edit the commit or give me a sign and I will clean it up.

wolandr commented:
Thanks for your feedback Christian! Using the `#![feature(alloc)]` seems really invalid for stable channel, so the request LGTM. But please make sure that the tests passing done with the changes. Probably the [feature choosing](https://github.com/wolandr/pretty-hex/blob/master/tests/tests.rs#L2) should be also changed in tests.rs.

sirhcel commented:
> But please make sure that the tests passing done with the changes.

The tests are up and running again when the _alloc_ feature is selected. Without, the previous test cases are all disabled because they make significant usage of features from _alloc_ for the tests itselves. Porting them to build without _alloc_ looks like a step back with respect to readability. So I left them as they are and added just one more test case for `hex_write` which works without _alloc_.

The existing coverage could be achieved by running the tests in both configurations:
```
$ cargo test --no-default-features
$ cargo test
```
This does not feel great. But sacrificing their readability seems even worse to me and I've got no better idea so far.

On this way I learned that Rust doc code gets tested as well. Pretty cool! But I found no way of conditionally generating the examples only when the _alloc_ feature is selected. So I went for disabling the `assert`s in the docs as well and hiding this in the generated docs. Do you know how I could achieve this?


> Probably the [feature choosing](https://github.com/wolandr/pretty-hex/blob/master/tests/tests.rs#L2) should be also changed in tests.rs.

Switched to `#[cfg(feature = "alloc")` here as well.

sirhcel commented:
I managed to mess up the commit message and cleaned it up.

wolandr commented:
I don't found a way to conditionally testing docs too, maybe something around `#![doc(...)]` macros, needs to investigate. However it looks good enough at this point. Thank you for contributing!
